VersionControlUI/src/nodes/DiffComparisonPair.cpp

```diff
@@ -25,14 +25,99 @@
  **********************************************************************************************************************/
 #include "DiffComparisonPair.h"
 
+#include "ModelBase/src/nodes/composite/CompositeNode.h"
+
 namespace VersionControlUI
 {
 
 DEFINE_NODE_TYPE_REGISTRATION_METHODS(DiffComparisonPair)
 
+DiffComparisonPair::DiffComparisonPair(Model::Node* oldVersionNode, Model::Node* newVersionNode) : Super{}
+{
+	oldVersionNode_ = oldVersionNode;
+	newVersionNode_ = newVersionNode;
+
+	setUpObjectPath();
+
+	setUpComponentType();
+
+}
+
+void DiffComparisonPair::setUpObjectPath()
+{
+	QString oldVersionObjectPath = computeObjectPath(oldVersionNode_);
+	QString newVersionObjectPath = computeObjectPath(newVersionNode_);
+	if (oldVersionObjectPath == newVersionObjectPath)
+	{
+		singleObjectPath_ = new Model::Text{oldVersionObjectPath};
+		twoObjectPathsDefined_ = false;
+	}
+	else if (oldVersionObjectPath == "" || newVersionObjectPath == "")
+	{
+		auto objectPath = oldVersionObjectPath == "" ? newVersionObjectPath : oldVersionObjectPath;
+		singleObjectPath_ = new Model::Text{objectPath};
+		twoObjectPathsDefined_ = false;
+	}
+	else
+	{
+		oldVersionObjectPath_ = new Model::Text{oldVersionObjectPath};
+		newVersionObjectPath_ = new Model::Text{newVersionObjectPath};
+		twoObjectPathsDefined_ = true;
+	}
+}
+
+QString DiffComparisonPair::computeObjectPath(Model::Node* node)
+{
+	if (!node) return "";
+	auto parent = node->parent();
+	QString objectPath = "";
+	while (parent)
+	{
+		if (parent->definesSymbol())
+			objectPath.prepend(parent->symbolName() + "/");
+		parent = parent->parent();
+	}
+
+	return objectPath;
+}
+
+void DiffComparisonPair::setUpComponentType()
+{
+	Model::Node* nodeToComputeComponentType = nullptr;
+	QString componentType = "unknown";
+
+	if (oldVersionNode_)
+		nodeToComputeComponentType = oldVersionNode_;
+	else
+		nodeToComputeComponentType = newVersionNode_;
+
+	auto searchCompositeNode = nodeToComputeComponentType;
+
+	while (searchCompositeNode && !DCast<Model::CompositeNode>(searchCompositeNode))
+		searchCompositeNode = searchCompositeNode->parent();
+
+	if (searchCompositeNode)
+	{
+		Model::CompositeNode* compositeNode = DCast<Model::CompositeNode>(searchCompositeNode);
+		if (compositeNode == nodeToComputeComponentType)
+		{
+			componentType = compositeNode->typeName();
+		} else
+		{
+			auto index = compositeNode->indexOf(nodeToComputeComponentType);
+			componentType = compositeNode->meta().attribute(index).name();
+		}
+
+	}
+	// TODO maybe first char of componentType toUpper?
+	componentType_ = new Model::Text{componentType};
+}
+
+// TODO force top constructor, is this right way? or remove macro which introduces the constructor?
```

Comment:
Ideally we don't have to write this dummy method, but if removing the macro means that we would have to write a lot of other things by hand, I'd rather keep the macro and this method.